app/controllers/charges_controller.rb

```diff
@@ -5,6 +5,13 @@
 class ChargesController < ApplicationController
   # POST /charges
   def create
+    # Validate this not a duplicate charge
+    idempotency_key = charge_params[:idempotency_key]
+    existing_event = ExistingEvent.new(idempotency_key: idempotency_key, event_type: 'charges_create')
+    unless existing_event.save
```

Comment:
Mind adding a unit test to the charges_spec here?

Reply:
Kinda struggling with...this is the test im envisioning, but it doesn't pass...not sure whats the problem
```
    context 'with a duplicate payment request' do
      let(:line_items) do
        [
          {
            amount: 50,
            currency: 'usd',
            item_type: 'gift_card',
            quantity: 1,
            seller_id: seller_id
          }
        ]
      end

      before { post '/charges', params: params, as: :json }

      it 'records charge events' do
        allow_any_instance_of(ExistingEvent).to receive(:save).and_return(false)
        expect(response).to have_http_status(400)
      end
    end
```

Reply:
Done